compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPULowerToUKernels.cpp

```diff
@@ -0,0 +1,261 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/Dialect/IREECodegenDialect.h"
+#include "iree/compiler/Codegen/Dialect/LoweringConfig.h"
+#include "iree/compiler/Codegen/Dialect/UKernelOps.h"
+#include "iree/compiler/Codegen/PassDetail.h"
+#include "iree/compiler/Codegen/Passes.h"
+#include "iree/compiler/Dialect/Flow/IR/FlowOps.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/Dialect/Bufferization/IR/Bufferization.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/IR/Diagnostics.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#define DEBUG_TYPE "iree-llvmgpu-lower-to-ukernels"
+
+namespace mlir {
+namespace iree_compiler {
+namespace {
+
+struct LLVMGPULowerToUKernelsPass
+    : LLVMGPULowerToUKernelsBase<LLVMGPULowerToUKernelsPass> {
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<IREE::Codegen::IREECodegenDialect>();
+    registry.insert<mlir::bufferization::BufferizationDialect>();
+  }
+  void runOnOperation() override;
+};
+
+static void generateMicrokernelName(std::string &fnName,
```

Comment:
I am not sure the call should be so tuned the tile size... The tile size should be passed as arguments, and after inlining this should all become a no-op.....

Reply:
While I understand that tiles sizes are eventually inlined, I wanted to clarify that we compile microkernels separately, and inlining happens after we linked them in the kernel. Tile sizes are needed in the compile-time while compiling microkernels.

Moreover, tile sizes are a template parameter in Cutlass classes. I cannot pass a function argument to them.

Reply:
Ok, I understand the issue here, but I am concerned about the blow up in the number of "microkernels" that will have to be used... So this doesnt seem to get to a reasonable end-state... The cost of maintaining and shipping so many variants just on tile sizes is non-trivial. I guess the decision isnt upto me... We should escalate this to someone who can be the decision maker on this.

Reply:
We identify the most commonly used tile sizes, approximately 5, and generating microkernels for those sizes are sufficient. I mean we don't need every tile size.

Also, generating microkernels with different tile sizes is trivial. I've implemented generator that generates microkernel by adding single line.